### PR TITLE
feat(validation): Add @validator decorator for JSON Schema validation

### LINKED ISSUE #3608: Feature request: `validator` class method decorator for JSON Schema validation
### Use case

To continue the implementation of the Validation utility started in #3607, we should implement a `@validator` decorator that can be used by TypeScript customers to wrap a class method and validate both inputs and outputs of the method against a set of JSON Schemas.

### Solution/User Experience

The decorator will use the `validate` function implemented in #3607 under the hood, and as such it should accept similar parameters to that function, albeit with some changes to accommodate the fact that **1/** it's a decorator, and **2/** it can validate both input and outputs of the decorated method.

The decorator will take a single `params` object that has the following items:
- an optional `inboundSchema` - this is the JSON Schema used to validate the incoming payload
- an optional `outboundSchema` - this is the JSON Schema used to validate the response payload
- an optional `envelope` parameter of type `string` (more on this later)
- an optional `formats` optional parameter to pass custom formats
- an optional `externalRefs` array of schemas that can be used to supply external references to the main schema
- an optional `ajv` parameter that customers can use to provide their own instance - this is useful when wanting to use a different JSON Schema version/draft

Both `inboundSchema` and `outboundSchema` are optional, this is intentional because we want to allow customers to validate one, the other, or both. If none is provided, the decorator will be a no-op.

In terms of implementation, the `validator` decorator will have the overall flow:
- if neither `inboundSchema` nor `outboundSchema` are provided, return and call the decorated method
- if `inboundSchema` is present, call the `validate` function using the schema as `schema` and the first argument of the decorated method as `payload`, as well as all the other optional params to validate the incoming payload
- call the decorated method using the result of the first `validate` call as first argument and store the `response` in a variable
- if `outboundSchema ` is present, call the `validate` function using the schema as `schema` and the `response` of the decorated method as `payload`, as well as all the other optional params except for `envelope` to validate it

Finally, since decorators cannot change the type of the parameters of the decorated function, there's no need for it to be a generic function nor an assertion function.

As part of the implementation the implementer should also write unit tests to cover 100% of the files, lines, branches of the code added as part of this issue. All the dependencies needed should already be present in the workspace, ideally the implementation should follow existing project conventions as much as possible. The decorator should be in its own file, i.e. `packages/validation/src/decorator.ts`.

When it comes to references, you can find an example of how this decorator would be used in the `docs/utilities/validation.md` file.

If instead you're unfamiliar with how to implement decorators, I've added an example below:

<details>

```ts
import type { Context, Handler } from 'aws-lambda';

export const validator = (
  params: ValidatorOptions
): HandlerMethodDecorator => {
  return (_target, _propertyKey, descriptor) => {
    // biome-ignore lint/style/noNonNullAssertion: The descriptor.value is the method this decorator decorates, it cannot be undefined.
    const original = descriptor.value!;

    // if no inbound nor outbound schemas are provided, return `descriptor` here

    descriptor.value = async function (
      this: Handler,
      event: unknown,
      context: Context,
      callback
    ) {
      // validate inbound schema here, if present
      const validatedEvent = '...';

      const response = original.call(this, validatedEvent, context, callback);

      // validate outbound schema here, if present
      return response;
    };

    return descriptor;
  };
};
```

</details>

### Alternative solutions

```Markdown
N/A
```

### Acknowledgment

- [x] This feature request meets [Powertools for AWS Lambda (TypeScript) Tenets](https://docs.powertools.aws.dev/lambda/typescript/latest/#tenets)
- [ ] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Python](https://github.com/aws-powertools/powertools-lambda-python/), [Java](https://github.com/aws-powertools/powertools-lambda-java/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### Future readers

Please react with 👍 and your use case to help us understand customer demand.

### PR DESCRIPTION
**Pull request by VatsalGoel3**
Changes

This PR introduces a @validator method decorator that enables JSON Schema validation for both input (inbound) and output (outbound) parameters of class methods. It is built using the existing validate function and follows the Powertools for AWS Lambda (TypeScript) validation utility standards.

Key Features
✅ Validates method arguments using inbound schemas
✅ Validates method returns values using outbound schemas
✅ Supports optional parameters like envelope, formats, and external references
✅ Allows custom AJV instances for extended schema validation
✅ Includes unit tests with 100% coverage

Implementation
Created validator.ts, a TypeScript decorator that validates the input and output of methods based on JSON Schema definitions.
Integrated the decorator with the existing validate function to ensure consistency.
Added unit tests to verify:
Successful validation for both inbound and outbound payloads.
Errors for invalid inbound and outbound payloads.
Proper handling when no schema is provided (no-op behavior).
Coverage for edge cases such as undefined method descriptors.

Issue Number
Closes #3608 (Feature request: Validator class method decorator for JSON Schema validation)

By submitting this pull request, I confirm that I have read and followed the Powertools for AWS Lambda (TypeScript) contributing guidelines. I confirm that you can use, modify, copy, and redistribute this contribution under the terms of your choice.

### CODE REVIEW

**Comment by dreamorosi** [comment]
Hi @VatsalGoel3 - thank you for opening the PR.

You have modified the PR body and our automation fails because of it. Not sure why this keeps happening, but if you're using some kind of git client, please set it up so it keeps the PR body intact and just fills in the required fields.

File: packages/validation/src/decorator.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+          });
+        } catch (error) {
+          throw new SchemaValidationError('Outbound Validation failed', error);
+        }
+      }
+      return result;
+    };
+    return descriptor;
+  };
+}
```
Comment (dreamorosi) [inline packages/validation/src/decorator.ts:79]:
```suggestion
import { SchemaValidationError } from './errors.js';
import { validate } from './validate.js';
import type { ValidatorOptions } from './types.js';

export function validator(options: ValidatorOptions) {
  return (
    _target: unknown,
    _propertyKey: string | symbol,
    descriptor: PropertyDescriptor
  ) => {
    if (!descriptor.value) {
      return descriptor;
    }

    const {
      inboundSchema,
      outboundSchema,
      envelope,
      formats,
      externalRefs,
      ajv,
    } = options;

    if (!inboundSchema && !outboundSchema) {
      return descriptor;
    }

    const originalMethod = descriptor.value;

    descriptor.value = async function (...args: unknown[]) {
      let validatedInput = args[0];

      if (inboundSchema) {
        try {
          validatedInput = validate({
            payload: validatedInput,
            schema: inboundSchema,
            envelope: envelope,
            formats: formats,
            externalRefs: externalRefs,
            ajv: ajv,
          });
        } catch (error) {
          throw new SchemaValidationError('Inbound validation failed', error);
        }
      }

      const result = await originalMethod.apply(this, [
        validatedInput,
        ...args.slice(1),
      ]);
      if (outboundSchema) {
        try {
          return validate({
            payload: result,
            schema: outboundSchema,
            formats: formats,
            externalRefs: externalRefs,
            ajv: ajv,
          });
        } catch (error) {
          throw new SchemaValidationError('Outbound Validation failed', error);
        }
      }
      return result;
    };
    return descriptor;
  };
}
```

File: packages/validation/src/decorator.ts
Comment (dreamorosi) [inline packages/validation/src/decorator.ts:1]: For the types, in the `types.ts` file, use these:

```ts
import type {
  Ajv,
  AsyncFormatDefinition,
  FormatDefinition,
  AnySchema,
} from 'ajv';

type Prettify<T> = {
  [K in keyof T]: T[K];
} & {};

type ValidateParams = {
  payload: unknown;
  schema: AnySchema;
  envelope?: string;
  formats?: Record<
    string,
    | string
    | RegExp
    | FormatDefinition<string>
    | FormatDefinition<number>
    | AsyncFormatDefinition<string>
    | AsyncFormatDefinition<number>
  >;
  externalRefs?: object[];
  ajv?: Ajv;
};

type ValidatorOptions = Prettify<
  Omit<ValidateParams, 'payload' | 'schema'> & {
    inboundSchema?: AnySchema;
    outboundSchema?: AnySchema;
  }
>;

export type { ValidateParams, ValidatorOptions };
```

**Review by dreamorosi — CHANGES_REQUESTED** [review]
Left a couple comments on the implementation  - mainly refactoring for types, the flow is good.

File: packages/validation/src/decorator.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+          });
+        } catch (error) {
+          throw new SchemaValidationError('Outbound validation failed', error);
+        }
+      }
+      return result;
+    };
+    return descriptor;
+  };
+}
```
Comment (dreamorosi) [inline packages/validation/src/decorator.ts:57]:
```suggestion
import { SchemaValidationError } from './errors.js';
import { validate } from './validate.js';
import type { ValidatorOptions } from './types.js';
export function validator(options: ValidatorOptions) {
  return (
    _target: unknown,
    _propertyKey: string | symbol,
    descriptor: PropertyDescriptor
  ) => {
    if (!descriptor.value) {
      return descriptor;
    }
    const {
      inboundSchema,
      outboundSchema,
      envelope,
      formats,
      externalRefs,
      ajv,
    } = options;
    if (!inboundSchema && !outboundSchema) {
      return descriptor;
    }
    const originalMethod = descriptor.value;
    descriptor.value = async function (...args: unknown[]) {
      let validatedInput = args[0];
      if (inboundSchema) {
        try {
          validatedInput = validate({
            payload: validatedInput,
            schema: inboundSchema,
            envelope: envelope,
            formats: formats,
            externalRefs: externalRefs,
            ajv: ajv,
          });
        } catch (error) {
          throw new SchemaValidationError('Inbound validation failed', error);
        }
      }
      const result = await originalMethod.apply(this, [
        validatedInput,
        ...args.slice(1),
      ]);
      if (outboundSchema) {
        try {
          return validate({
            payload: result,
            schema: outboundSchema,
            formats: formats,
            externalRefs: externalRefs,
            ajv: ajv,
          });
        } catch (error) {
          throw new SchemaValidationError('Outbound Validation failed', error);
        }
      }
      return result;
    };
    return descriptor;
  };
}
```

**Review by dreamorosi — CHANGES_REQUESTED** [review]
Thanks for the review, before merging we need to make a few more changes:
- In `packages/validation/src/validate.ts` change the import from `import Ajv, { type ValidateFunction } from 'ajv';` to `import { Ajv, type ValidateFunction } from 'ajv';`
- In `packages/validation/src/validate.ts` change the function signature from `export function validate<T = unknown>(params: ValidateParams<T>): T {` to `export function validate<T = unknown>(params: ValidateParams): T {`
- In `packages/validation/src/index.ts` change the first import from `export { validate } from './validate';` to `export { validate } from './validate.js';`
- In `packages/validation/src/index.ts` add a new import `export { validator } from './decorator.js';`

I checked out the branch and tried to build it, and unless we make these changes it won't work.

**Comment by VatsalGoel3** [comment]
@dreamorosi Just updated with the suggested changes.

**Review by dreamorosi — APPROVED** [review]
Thank you for addressing all the review comments!

Let's ship it 🚢 !
